org.dom4j.DocumentException on resolving schema

**Describe the bug**
If XML is not well formed, but has `schemaLocation` defined in the root tag, it fails with following error.
This is because, in order to resolve schemas, it first tries to parse it and then tries to show the error. This should be avoided 
_i.e. if XML is malformed, first display those error and then if its not then display schema validations._
```
17:00:50.377 [pool-2-thread-1] ERROR io.github.handofgod94.schema.resolve.XsdSchemaResolver - Unable to parse or load schema
org.dom4j.DocumentException: Error on line 18 of document  : The element type "choice" must be terminated by the matching end-tag "</choice>".
	at org.dom4j.io.SAXReader.read(SAXReader.java:462) ~[xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at org.dom4j.io.SAXReader.read(SAXReader.java:345) ~[xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.github.handofgod94.schema.resolve.XsdSchemaResolver.resolve(XsdSchemaResolver.java:44) ~[xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.github.handofgod94.main.XmlDocumentService.didSave(XmlDocumentService.java:145) ~[xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_191]
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_191]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_191]
	at java.lang.reflect.Method.invoke(Unknown Source) ~[?:1.8.0_191]
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65) ~[xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.notify(GenericEndpoint.java:152) [xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleNotification(RemoteEndpoint.java:220) [xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:187) [xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:192) [xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94) [xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:99) [xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:1.8.0_191]
	at java.util.concurrent.FutureTask.run(Unknown Source) [?:1.8.0_191]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:1.8.0_191]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:1.8.0_191]
	at java.lang.Thread.run(Unknown Source) [?:1.8.0_191]
Caused by: org.xml.sax.SAXParseException: The element type "choice" must be terminated by the matching end-tag "</choice>".
	at org.apache.xerces.util.ErrorHandlerWrapper.createSAXParseException(Unknown Source) ~[xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at org.apache.xerces.util.ErrorHandlerWrapper.fatalError(Unknown Source) ~[xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at org.apache.xerces.impl.XMLErrorReporter.reportError(Unknown Source) ~[xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at org.apache.xerces.impl.XMLErrorReporter.reportError(Unknown Source) ~[xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at org.apache.xerces.impl.XMLErrorReporter.reportError(Unknown Source) ~[xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at org.apache.xerces.impl.XMLScanner.reportFatalError(Unknown Source) ~[xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at org.apache.xerces.impl.XMLNSDocumentScannerImpl.scanEndElement(Unknown Source) ~[xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl$FragmentContentDispatcher.dispatch(Unknown Source) ~[xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl.scanDocument(Unknown Source) ~[xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source) ~[xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source) ~[xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at org.apache.xerces.parsers.XMLParser.parse(Unknown Source) ~[xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at org.apache.xerces.parsers.AbstractSAXParser.parse(Unknown Source) ~[xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at org.apache.xerces.jaxp.SAXParserImpl$JAXPSAXParser.parse(Unknown Source) ~[xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at org.dom4j.io.SAXReader.read(SAXReader.java:445) ~[xml-language-server-1.0-SNAPSHOT-jar-with-dependencies.jar:?]
	... 19 more
```

**To Reproduce**

Steps to reproduce the behavior:
1. Open any well formed XML File with valid `schemaLocation`
2. Add an extra element but just the starting tag and not the ending tag.
3. Save the file.

**Expected behavior**
It should first display malformed XML error in diagnostic report.

**Desktop (please complete the following information):**
 - OS : Windows
 - Editor :  VSCode
 - Editor Version : 1.30.1

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -184,7 +184,7 @@ endpoints defined in [language server specification](https://microsoft.github.io
 
 6. The current remote debugging port for Java is `8010`. You can connect from any ide
    to this port for Remote debugging.  
-   If you want to change it then it can be modified in `xml-vscode-clien/src/extension.ts` file
+   If you want to change it then it can be modified in `xml-vscode-plugin/src/extension.ts` file
    ```typescript
    // Server options
     let serverOpts: ServerOptions = {

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -204,7 +204,7 @@ endpoints defined in [language server specification](https://microsoft.github.io
                 '-Djava.util.logging.manager=org.apache.logging.log4j.jul.LogManager',
                 '-jar',
                 
-                //Modfify the address to change the debugging port
+                //Modify the address to change the debugging port
                 '-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,quiet=y,address=8010',
                 debugXmlServerPath
             ],

--- a/src/main/java/io/github/handofgod94/common/XmlUtil.java
+++ b/src/main/java/io/github/handofgod94/common/XmlUtil.java
@@ -1,5 +1,6 @@
 package io.github.handofgod94.common;
 
+import io.github.handofgod94.common.parser.CustomErrorHandler;
 import io.github.handofgod94.common.parser.PositionalHandler;
 import java.io.ByteArrayInputStream;
 import java.io.IOException;
@@ -9,12 +10,18 @@ import java.nio.charset.StandardCharsets;
 import java.util.Arrays;
 import java.util.HashSet;
 import java.util.List;
+import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
+import javax.xml.XMLConstants;
 import javax.xml.namespace.QName;
 import javax.xml.parsers.ParserConfigurationException;
 import javax.xml.parsers.SAXParser;
 import javax.xml.parsers.SAXParserFactory;
+import javax.xml.transform.stream.StreamSource;
+import javax.xml.validation.Schema;
+import javax.xml.validation.SchemaFactory;
+import javax.xml.validation.Validator;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.apache.xerces.xs.XSConstants;
@@ -26,6 +33,7 @@ import org.apache.xerces.xs.XSParticle;
 import org.dom4j.Document;
 import org.dom4j.DocumentException;
 import org.dom4j.io.SAXReader;
+import org.eclipse.lsp4j.Position;
 import org.xml.sax.SAXException;
 
 /**
@@ -203,5 +211,46 @@ public class XmlUtil {
       }
     }
     return Optional.empty();
+  }
+
+  /**
+   * Generates XML Schema instance object.
+   * @apiNote  Since we are not providing any sources for schema, then it'll
+   *     automatically look for schema location in the document itself and will
+   *     load the schema.
+   * @return A standard XML Schema object following 2001 XML namespace
+   * @throws SAXException If unable to create schema object.
+   */
+  public static Schema generateSchema() throws SAXException {
+    // Generate schema using the stream sources
+    SchemaFactory factory = SchemaFactory.newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI);
+    Schema schema = factory.newSchema();
+    return schema;
+  }
+
+  /**
+   * Check for well formed XML.
+   * The schema validation and other LSP features, could only work with
+   * well formed XML documents. This utility provides a helper function
+   * to check for XML errors. If any errors are present, then it'll calculate
+   * position and error messages present in the XML and will store it in a map.
+   * If no errors are present, then it'll return an empty Map.
+   * @param text xml text which needs to be checked
+   * @return A Map containing error position and message if present, empty otherwise.
+   */
+  public static Map<Position, String> checkWellFormedXml(String text) {
+    CustomErrorHandler errorHandler = new CustomErrorHandler();
+    try {
+      Validator validator = generateSchema().newValidator();
+      validator.setErrorHandler(errorHandler);
+
+      validator.validate(new StreamSource(new StringReader(text)));
+    } catch (SAXException | IOException e) {
+      logger.info("Document seems to be malformed or unable to retrieve schema", e);
+    }
+
+    Map<Position, String> errorMap = errorHandler.getErrorMap();
+
+    return errorMap;
   }
 }

--- a/src/main/java/io/github/handofgod94/common/document/DocumentManager.java
+++ b/src/main/java/io/github/handofgod94/common/document/DocumentManager.java
@@ -35,7 +35,7 @@ public class DocumentManager {
   private String[] documentLines = null;
 
   @Inject
-  public DocumentManager(@Assisted TextDocumentItem documentItem) {
+  DocumentManager(@Assisted TextDocumentItem documentItem) {
     this.documentItem = documentItem;
     this.documentLines = XmlUtil.getDocumentLines(this.documentItem.getText());
   }

--- a/src/main/java/io/github/handofgod94/common/parser/CustomErrorHandler.java
+++ b/src/main/java/io/github/handofgod94/common/parser/CustomErrorHandler.java
@@ -1,4 +1,4 @@
-package io.github.handofgod94.lsp.diagnostic;
+package io.github.handofgod94.common.parser;
 
 import java.util.HashMap;
 import java.util.Map;
@@ -13,7 +13,7 @@ import org.xml.sax.SAXParseException;
  * xml text. This class extends the base ErrorHandler class and add all the
  * errors and warnings to the list of errors.
  */
-public class DiagnosticErrorHandler implements ErrorHandler {
+public class CustomErrorHandler implements ErrorHandler {
 
   // Error map to store line number and errors description and severity.
   private Map<Position, String> errorMap = new HashMap<>();

--- a/src/main/java/io/github/handofgod94/grammar/GrammarProcessor.java
+++ b/src/main/java/io/github/handofgod94/grammar/GrammarProcessor.java
@@ -19,7 +19,7 @@ import org.eclipse.lsp4j.Position;
 public class GrammarProcessor {
 
   /**
-   * Guice factory for GrammarProcessor
+   * Guice factory for GrammarProcessor.
    */
   public interface Factory {
     GrammarProcessor create(Position position, String line);

--- a/src/main/java/io/github/handofgod94/grammar/GrammarProcessor.java
+++ b/src/main/java/io/github/handofgod94/grammar/GrammarProcessor.java
@@ -18,6 +18,9 @@ import org.eclipse.lsp4j.Position;
  */
 public class GrammarProcessor {
 
+  /**
+   * Guice factory for GrammarProcessor
+   */
   public interface Factory {
     GrammarProcessor create(Position position, String line);
   }

--- a/src/main/java/io/github/handofgod94/lsp/diagnostic/XmlDiagnosticService.java
+++ b/src/main/java/io/github/handofgod94/lsp/diagnostic/XmlDiagnosticService.java
@@ -1,112 +1,56 @@
 package io.github.handofgod94.lsp.diagnostic;
 
-import com.google.inject.Inject;
-import com.google.inject.assistedinject.Assisted;
-import io.github.handofgod94.common.document.DocumentManager;
-import io.github.handofgod94.main.XmlLanguageServer;
-import io.github.handofgod94.schema.SchemaDocument;
-import java.io.IOException;
-import java.io.StringReader;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
-import javax.xml.transform.stream.StreamSource;
-import javax.xml.validation.Validator;
-import org.apache.logging.log4j.LogManager;
-import org.apache.logging.log4j.Logger;
 import org.eclipse.lsp4j.Diagnostic;
 import org.eclipse.lsp4j.DiagnosticSeverity;
 import org.eclipse.lsp4j.Position;
 import org.eclipse.lsp4j.PublishDiagnosticsParams;
 import org.eclipse.lsp4j.Range;
-import org.eclipse.lsp4j.TextDocumentItem;
-import org.xml.sax.SAXException;
 
 public class XmlDiagnosticService {
 
   /**
-   * Factory for XmlDiagnosticService.
-   * This is used for injecting {@link XmlDiagnosticService} instance.
+   * Get parameters for diagnostic, based on the error map.
+   * The error map should consist the position of error in the document
+   * as well as message for that error. These params can be published to client
+   * using XML Language server object
+   * @param lines array of document split line by line
+   * @param documentUri uri of the document for which diagnostics are requested
+   * @param errorMap HashMap consisting of position of error and the error message
+   * @return PublishDiagnosticsParams instance having all the diagnostics.
    */
-  public interface Factory {
-    /**
-     * Factory method to create service.
-     * @param documentItem currently working text item
-     * @param server XmlLanguageServer object
-     * @param schemaDocument Instance of schema document
-     * @return XmlDiagnosticService instance
-     * @see io.github.handofgod94.schema.SchemaDocument
-     */
-    XmlDiagnosticService create(TextDocumentItem documentItem,
-                                XmlLanguageServer server,
-                                SchemaDocument schemaDocument);
-  }
-
-  private static final Logger logger = LogManager.getLogger(XmlDiagnosticService.class.getName());
-
-  private SchemaDocument schemaDocument;
-  private XmlLanguageServer server;
-  private TextDocumentItem documentItem;
-  private DocumentManager.Factory documentManagerFactory;
-  private DiagnosticErrorHandler errorHandler;
-
-  @Inject
-  XmlDiagnosticService(@Assisted TextDocumentItem documentItem, @Assisted XmlLanguageServer server,
-                       @Assisted SchemaDocument schemaDocument,
-                       DocumentManager.Factory documentManagerFactory,
-                       DiagnosticErrorHandler errorHandler) {
-    this.documentItem = documentItem;
-    this.server = server;
-    this.schemaDocument = schemaDocument;
-    this.documentManagerFactory = documentManagerFactory;
-    this.errorHandler = errorHandler;
-  }
-
-  /**
-   * Gets all the errors in the XML file and push it to client. This needs to
-   * invoked on load/save of document.
-   */
-  public void compute() {
-    String text = documentItem.getText();
-    Validator validator = schemaDocument.getSchema().newValidator();
-    validator.setErrorHandler(errorHandler);
-
-    try {
-      validator.validate(new StreamSource(new StringReader(text)));
-    } catch (SAXException | IOException ex) {
-      logger.error("Error occurred while parsing/reading the document", ex);
-    }
-
-    // Error Map
-    // This will be calculated in DiagnosticErrorHandler class.
-    Map<Position, String> errorMap = errorHandler.getErrorMap();
-
+  public static PublishDiagnosticsParams getDiagnostics(String[] lines,
+                                                        String documentUri,
+                                                        Map<Position, String> errorMap) {
     // Create diagnostic collection and getCompletionItems data from error map
-    List<Diagnostic> diagnostics = new ArrayList<>(getDiagnosticsFromErrorMap(errorMap));
+    List<Diagnostic> diagnostics = new ArrayList<>(getDiagnosticsFromErrorMap(lines, errorMap));
 
     // Publish diagnostics to client
     PublishDiagnosticsParams diagnosticsParams =
-        new PublishDiagnosticsParams(this.documentItem.getUri(), diagnostics);
-    server.getClient().publishDiagnostics(diagnosticsParams);
+        new PublishDiagnosticsParams(documentUri, diagnostics);
+
+    return diagnosticsParams;
   }
 
   /**
    * Traverse errorMap and create list of diagnostic objects.
    *
+   * @param lines array of string having content of document line by line.
+   * @param errorMap Map of position to error obtained from Xml Document
    * @return List of diagnostics
    */
-  protected List<Diagnostic> getDiagnosticsFromErrorMap(Map<Position, String> errorMap) {
+  private static List<Diagnostic> getDiagnosticsFromErrorMap(String[] lines,
+                                                             Map<Position, String> errorMap) {
     // Create diagnostic collection
     List<Diagnostic> diagnostics = new ArrayList<>();
-
-    // Create DocumentManager for querying text document
-    DocumentManager manager = documentManagerFactory.create(documentItem);
 
     for (Position position : errorMap.keySet()) {
       int lineNo = position.getLine();
       int endColumn = position.getCharacter();
       String message = errorMap.get(position);
-      String line = manager.getDocumentLines()[lineNo];
+      String line = lines[lineNo];
       int startColumn = line.length() - line.trim().length();
 
       Position startPosition = new Position(lineNo, startColumn);

--- a/src/main/java/io/github/handofgod94/main/XmlDocumentService.java
+++ b/src/main/java/io/github/handofgod94/main/XmlDocumentService.java
@@ -100,6 +100,8 @@ public class XmlDocumentService implements TextDocumentService {
     return CompletableFuture.completedFuture(Either.forLeft(list));
   }
 
+
+
   @Override
   public void didOpen(DidOpenTextDocumentParams params) {
     logger.info("File opened: {}", params.getTextDocument().getUri());
@@ -110,7 +112,7 @@ public class XmlDocumentService implements TextDocumentService {
     Map<Position, String> errorMessages = XmlUtil.checkWellFormedXml(documentItem.getText());
     publishDiagnostics(documentItem, errorMessages);
 
-    if (errorMessages.isEmpty()) {
+    if (errorMessages.isEmpty() && schemaDocument == null) {
       // Load XSD Schema model
       Optional<SchemaDocument> schemaDocumentOptional = resolver.resolve(documentItem.getText());
       schemaDocumentOptional.ifPresent(document -> schemaDocument = document);
@@ -143,7 +145,7 @@ public class XmlDocumentService implements TextDocumentService {
     Map<Position, String> errorMessages = XmlUtil.checkWellFormedXml(documentItem.getText());
     publishDiagnostics(documentItem, errorMessages);
 
-    if (errorMessages.isEmpty()) {
+    if (errorMessages.isEmpty() && schemaDocument == null) {
       // Load XSD Schema model
       Optional<SchemaDocument> schemaDocumentOptional = resolver.resolve(documentItem.getText());
       schemaDocumentOptional.ifPresent(document -> schemaDocument = document);

--- a/src/main/java/io/github/handofgod94/main/XmlDocumentService.java
+++ b/src/main/java/io/github/handofgod94/main/XmlDocumentService.java
@@ -3,6 +3,8 @@ package io.github.handofgod94.main;
 import com.google.inject.Inject;
 import com.google.inject.Key;
 import com.google.inject.name.Names;
+import io.github.handofgod94.common.XmlUtil;
+import io.github.handofgod94.common.document.DocumentManager;
 import io.github.handofgod94.lsp.completion.XmlCompletion;
 import io.github.handofgod94.lsp.completion.XmlCompletionFactory;
 import io.github.handofgod94.lsp.diagnostic.XmlDiagnosticService;
@@ -29,6 +31,7 @@ import org.eclipse.lsp4j.DidOpenTextDocumentParams;
 import org.eclipse.lsp4j.DidSaveTextDocumentParams;
 import org.eclipse.lsp4j.Hover;
 import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.PublishDiagnosticsParams;
 import org.eclipse.lsp4j.TextDocumentContentChangeEvent;
 import org.eclipse.lsp4j.TextDocumentItem;
 import org.eclipse.lsp4j.TextDocumentPositionParams;
@@ -46,20 +49,21 @@ public class XmlDocumentService implements TextDocumentService {
   private Map<String, TextDocumentItem> openDocumentItems = new HashMap<>();
   private SchemaDocument schemaDocument;
 
-  @Inject private final XmlDiagnosticService.Factory diagnosticServiceFactory;
   @Inject private final XmlCompletionFactory xmlCompletionFactory;
   @Inject private final XmlHoverProviderFactory xmlHoverProviderFactory;
+  @Inject private final DocumentManager.Factory docManagerFactory;
   @Inject private final SchemaResolver resolver;
 
   /**
    * Create Document service for XML documents.
+   *
    * @param server the language server instance
    */
   public XmlDocumentService(XmlLanguageServer server) {
     this.server = server;
-    diagnosticServiceFactory = server.getInjector().getInstance(XmlDiagnosticService.Factory.class);
     xmlHoverProviderFactory = server.getInjector().getInstance(XmlHoverProviderFactory.class);
     xmlCompletionFactory = server.getInjector().getInstance(XmlCompletionFactory.class);
+    docManagerFactory = server.getInjector().getInstance(DocumentManager.Factory.class);
     // TODO: named injection should be based on dtd or xsd texts.
     resolver = server.getInjector().getInstance(Key.get(SchemaResolver.class, Names.named("Xsd")));
   }
@@ -75,8 +79,8 @@ public class XmlDocumentService implements TextDocumentService {
 
     // Show hover if documentation is present, else show empty/no hover.
     Hover hover = optHover
-                  .map(XmlHover::getHover)
-                  .orElse(new Hover(Collections.emptyList()));
+        .map(XmlHover::getHover)
+        .orElse(new Hover(Collections.emptyList()));
     return CompletableFuture.completedFuture(hover);
   }
 
@@ -103,16 +107,14 @@ public class XmlDocumentService implements TextDocumentService {
     TextDocumentItem documentItem = params.getTextDocument();
     openDocumentItems.put(documentItem.getUri(), documentItem);
 
-    // Get instance from google guice injector
-    Optional<SchemaDocument> optSchemaDocument = resolver.resolve(documentItem.getText());
+    Map<Position, String> errorMessages = XmlUtil.checkWellFormedXml(documentItem.getText());
+    publishDiagnostics(documentItem, errorMessages);
 
-    // Generate diagnostics when xml document is opened
-    optSchemaDocument.ifPresent(schemaDocument -> {
-      this.schemaDocument = schemaDocument;
-      XmlDiagnosticService service =
-          diagnosticServiceFactory.create(documentItem, server, schemaDocument);
-      service.compute();
-    });
+    if (errorMessages.isEmpty()) {
+      // Load XSD Schema model
+      Optional<SchemaDocument> schemaDocumentOptional = resolver.resolve(documentItem.getText());
+      schemaDocumentOptional.ifPresent(document -> schemaDocument = document);
+    }
   }
 
   @Override
@@ -138,18 +140,21 @@ public class XmlDocumentService implements TextDocumentService {
     TextDocumentItem documentItem = openDocumentItems.get(params.getTextDocument().getUri());
     documentItem.setText(params.getText());
 
-    // TODO: This should be observable, if document changes, then it requires resync.
+    Map<Position, String> errorMessages = XmlUtil.checkWellFormedXml(documentItem.getText());
+    publishDiagnostics(documentItem, errorMessages);
 
-    // Currently, its again fetching the whole xsd and reparsing it
-    // to create new xs models. This must be avoided.
-    Optional<SchemaDocument> optSchemaDocument = resolver.resolve(documentItem.getText());
+    if (errorMessages.isEmpty()) {
+      // Load XSD Schema model
+      Optional<SchemaDocument> schemaDocumentOptional = resolver.resolve(documentItem.getText());
+      schemaDocumentOptional.ifPresent(document -> schemaDocument = document);
+    }
+  }
 
-    // Generate diagnostics on open of a text document
-    optSchemaDocument.ifPresent(schemaDocument -> {
-      this.schemaDocument = schemaDocument;
-      XmlDiagnosticService service =
-          diagnosticServiceFactory.create(documentItem, server, schemaDocument);
-      service.compute();
-    });
+  private void publishDiagnostics(TextDocumentItem documentItem, Map<Position, String> errorMap) {
+    DocumentManager manager = docManagerFactory.create(documentItem);
+    String[] lines = manager.getDocumentLines();
+    PublishDiagnosticsParams diagnostics =
+        XmlDiagnosticService.getDiagnostics(lines, documentItem.getUri(), errorMap);
+    server.getClient().publishDiagnostics(diagnostics);
   }
 }

--- a/src/main/java/io/github/handofgod94/main/XmlLanguageServerModule.java
+++ b/src/main/java/io/github/handofgod94/main/XmlLanguageServerModule.java
@@ -4,12 +4,11 @@ import com.google.inject.AbstractModule;
 import com.google.inject.assistedinject.FactoryModuleBuilder;
 import com.google.inject.name.Names;
 import io.github.handofgod94.common.document.DocumentManager;
+import io.github.handofgod94.common.parser.CustomErrorHandler;
 import io.github.handofgod94.common.parser.PositionalHandler;
 import io.github.handofgod94.grammar.GrammarProcessor;
 import io.github.handofgod94.grammar.graph.LanguageGraphContext;
 import io.github.handofgod94.lsp.completion.XmlCompletionFactory;
-import io.github.handofgod94.lsp.diagnostic.DiagnosticErrorHandler;
-import io.github.handofgod94.lsp.diagnostic.XmlDiagnosticService;
 import io.github.handofgod94.lsp.hover.AttributeHover;
 import io.github.handofgod94.lsp.hover.ElementHover;
 import io.github.handofgod94.lsp.hover.XmlHover;
@@ -32,14 +31,13 @@ public class XmlLanguageServerModule extends AbstractModule {
     bind(SchemaResolver.class).annotatedWith(Names.named("Xsd")).to(XsdSchemaResolver.class);
 
     // Bindings for concrete class
-    bind(DiagnosticErrorHandler.class);
+    bind(CustomErrorHandler.class);
     bind(XmlCompletionFactory.class);
     bind(LanguageGraphContext.class);
 
     // FactoryBuilders for assisted injections
     install(new FactoryModuleBuilder().build(DocumentManager.Factory.class));
     install(new FactoryModuleBuilder().build(PositionalHandler.Factory.class));
-    install(new FactoryModuleBuilder().build(XmlDiagnosticService.Factory.class));
     install(new FactoryModuleBuilder().build(XmlHoverProviderFactory.class));
     install(new FactoryModuleBuilder()
         .implement(XmlHover.class, Names.named("Element"), ElementHover.class)

--- a/src/main/java/io/github/handofgod94/schema/resolve/XsdSchemaResolver.java
+++ b/src/main/java/io/github/handofgod94/schema/resolve/XsdSchemaResolver.java
@@ -1,5 +1,6 @@
 package io.github.handofgod94.schema.resolve;
 
+import io.github.handofgod94.common.XmlUtil;
 import io.github.handofgod94.schema.SchemaDocument;
 import io.github.handofgod94.schema.SchemaDocumentType;
 import java.io.IOException;
@@ -8,10 +9,8 @@ import java.net.URI;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Optional;
-import javax.xml.XMLConstants;
 import javax.xml.transform.stream.StreamSource;
 import javax.xml.validation.Schema;
-import javax.xml.validation.SchemaFactory;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.apache.xerces.impl.xs.XMLSchemaLoader;
@@ -46,8 +45,7 @@ public class XsdSchemaResolver implements SchemaResolver {
       List<URI> schemaUris = searchSchemaUris(xmlDocument);
 
       // generate schema and models
-      StreamSource[] sources = generateSources(schemaUris);
-      Schema schema = generateSchema(sources);
+      Schema schema = XmlUtil.generateSchema();
       XSLoader xsLoader = new XMLSchemaLoader();
       XSModel xsModel = xsLoader.loadURIList(getUriStringList(schemaUris));
 
@@ -111,13 +109,6 @@ public class XsdSchemaResolver implements SchemaResolver {
       sources[i] = new StreamSource(schemaUris.get(i).toURL().openStream());
     }
     return sources;
-  }
-
-  private Schema generateSchema(StreamSource[] sources) throws SAXException {
-    // Generate schema using the stream sources
-    SchemaFactory factory = SchemaFactory.newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI);
-    Schema schema = factory.newSchema(sources);
-    return schema;
   }
 
   private StringList getUriStringList(List<URI> uris) {

